src/TRestGeant4ParticleSource.cxx

```diff
@@ -42,51 +47,316 @@ TRestGeant4ParticleSource::~TRestGeant4ParticleSource() {
+                 << GetDirection().Z() << ")" << endl;
+        metadata << "Energy distribution : " << GetEnergyDistType() << endl;
+        if (GetEnergyDistType() == "TH1D") {
+            metadata << "Energy distribution filename : " << GetSpectrumFilename() << endl;
+            metadata << "Energy histogram name  : " << GetSpectrumName() << endl;
+        }
+        if (GetEnergyRange().X() == GetEnergyRange().Y())
+            metadata << "Energy : " << GetEnergy() << " keV" << endl;
+        else
+            metadata << "Energy range : (" << GetEnergyRange().X() << "," << GetEnergyRange().Y() << ") keV"
+                     << endl;
     }
-    metadata << "Direction : (" << GetDirection().X() << "," << GetDirection().Y() << ","
-             << GetDirection().Z() << ")" << endl;
-    metadata << "Energy distribution : " << GetEnergyDistType() << endl;
-    if (GetEnergyDistType() == "TH1D") {
-        metadata << "Energy distribution filename : " << GetSpectrumFilename() << endl;
-        metadata << "Energy histogram name  : " << GetSpectrumName() << endl;
+}
+
+TRestGeant4ParticleSource* TRestGeant4ParticleSource::instantiate(std::string model) {
+    if (model == "" || model == "geant4" || model.find(".dat") != -1) {
+        // use default generator
+        return new TRestGeant4ParticleSource();
+    } else {
+        // use specific generator
+        // naming convension: TRestGeant4ParticleSourceXXX
+        // currently supported generator: decay0
+        // in future we may add wrapper of generators: cry(for muon), pythia(for HEP), etc.
+        model[0] = *REST_StringHelper::ToUpper(std::string(&model[0], 1)).c_str();
+        TClass* c = TClass::GetClass(("TRestGeant4ParticleSource" + model).c_str());
+        if (c != NULL)  // this means we have the package installed
+        {
+            return (TRestGeant4ParticleSource*)c->New();
+        } else {
+            std::cout << "REST ERROR! generator wrapper \"" << ("TRestGeant4ParticleSource" + model)
+                      << "\" not found!" << std::endl;
+        }
+    }
+    return NULL;
+}
+
+void TRestGeant4ParticleSource::InitFromConfigFile() {
+    TString modelUse = GetParameter("use");
+    if (((string)modelUse).find(".dat") != -1) {
+        string fullPathName = SearchFile((string)modelUse);
+        if (fullPathName == "") {
+            ferr << "File not found : " << modelUse << endl;
+            ferr << "Decay0 generator file could not be found!!" << endl;
+            exit(1);
+        }
+        modelUse = fullPathName;
+    }
+    SetGenFilename(modelUse);
+
+    if (((string)fGenFilename).find(".dat") != -1) {
+        if (TRestTools::fileExists((string)fGenFilename)) {
+            ReadEventDataFile(fGenFilename);
+        }
+    }
+}
+
+// base class's generator action: randomize the particle's energy/direction with distribution file
+void TRestGeant4ParticleSource::Update() {
+    if (fParticlesTemplate.size() > 0) {
```

Comment:
`if ( !fParticlesTemplate.empty() )`

Reply:
We use that `.size()` everywhere. I guess they are equivalent, and good enough to read and understand its meaning.